test_mavros/include/tests/offboard_control.h

```diff
@@ -0,0 +1,470 @@
+			case 5:
+				X = sp_X; Y = sp_Y; Z = sp_Z;
+				break;
+			default:
+				break;
+			}
+
+			if (pos_target == 6) {
+				ROS_INFO("Test complete!");
+				ros::shutdown();
+			}
+			else
+				++pos_target;
+
+			loop_rate.sleep();
+			ros::spinOnce();
+		}
+	}
+
+	/**
+	 * @brief Circle path motion routine
+	 */
+	void circle_path_motion(ros::Rate loop_rate){
+		local_pos_sp_pub = nh_sp.advertise<geometry_msgs::PoseStamped>("/mavros/setpoint_position/local", 10);
+		vel_sp_pub = nh_sp.advertise<geometry_msgs::TwistStamped>("/mavros/setpoint_velocity/cmd_vel", 10);
+		local_pos_sub = nh_sp.subscribe("/mavros/local_position/local", 10, &OffboardControl::local_pos_cb, this);
+
+		ROS_INFO("Testing...");
+
+		while (ros::ok()) {
+			// starting point
+			if (mode.compare("position") == 0) {
+				X = 5.0f;
+				Y = 0.0f;
+				Z = 1.0f;
+				local_pos_sp_pub.publish(ps);
+			}
+			else if (mode.compare("velocity") == 0) {
+				VX = 5.0f - current_x;
+				VY = 0.0f - current_y;
+				VZ = 1.0f - current_z;
+				vel_sp_pub.publish(vs);
+			}
+			else if (mode.compare("acceleration") == 0) {
+				// TODO
+				return;
+			}
+
+			wait_destination(ps);
+
+			// motion routine
+			for (int theta = 0; theta <= 360; theta++) {
+				if (mode.compare("position") == 0)
+					local_pos_sp_pub.publish(circle_shape(theta));
+				else if (mode.compare("velocity") == 0) {
+					VX = circle_shape(theta).pose.position.x - current_x;
+					VY = circle_shape(theta).pose.position.y - current_y;
+					VZ = circle_shape(theta).pose.position.z - current_z;
+					vel_sp_pub.publish(vs);
+				}
+				else if (mode.compare("acceleration") == 0) {
+					// TODO
+					return;
+				}
+				if (theta == 360) {
+					ROS_INFO("Test complete!");
+					ros::shutdown();
+				}
+				loop_rate.sleep();
+				ros::spinOnce();
+			}
+		}
+	}
+
+	/**
+	 * @brief Eight path motion routine
+	 */
+	void eight_path_motion(ros::Rate loop_rate){
+		local_pos_sp_pub = nh_sp.advertise<geometry_msgs::PoseStamped>("/mavros/setpoint_position/local", 10);
+		vel_sp_pub = nh_sp.advertise<geometry_msgs::TwistStamped>("/mavros/setpoint_velocity/cmd_vel", 10);
+		local_pos_sub = nh_sp.subscribe("/mavros/local_position/local", 10, &OffboardControl::local_pos_cb, this);
+
+		ROS_INFO("Testing...");
+
+		while (ros::ok()) {
+			// starting point
+			if (mode.compare("position") == 0) {
+				X = 0.0f;
+				Y = 0.0f;
+				Z = 1.0f;
+				local_pos_sp_pub.publish(ps);
+			}
+			else if (mode.compare("velocity") == 0) {
+				VX = 0.0f - current_x;
+				VY = 0.0f - current_y;
+				VZ = 1.0f - current_z;
+				vel_sp_pub.publish(vs);
+			}
+			else if (mode.compare("acceleration") == 0) {
+				// TODO
+				return;
+			}
+
+			wait_destination(ps);
+
+			// motion routine
+			for (int theta = -180; theta <= 180; theta++) {
+				if (mode.compare("position") == 0)
+					local_pos_sp_pub.publish(eight_shape(theta));
+				else if (mode.compare("velocity") == 0) {
+					VX = eight_shape(theta).pose.position.x - current_x;
```

Comment:
Oh ok, that's more concrete. So yes, store `eight_shape(theta);` at the beginning of the iteration and then use it in each calculation. Sounds good ;)